### DISCUSSION TITLE
How to commit pending changes on query invoked auto-flush instead of flush in SQLAlchemy?

### DISCUSSION THREAD

**Discussion by liveuser123** [sqlalchemy/sqlalchemy#9566] [Usage Questions]
I am using SQLAlchemy and I have a scenario where I want to commit any pending changes in the session when a query-invoked auto-flush is done, rather than just flushing them. I have tried adding event listeners for before_flush, but using session.commit() in before_flush did not persist the changes and using it in after_flush raised a ResourceClosedError. Here's an example of the event listeners I added:
```
@event.listens_for(Session, "before_flush")
def commit_pending_changes(session, flush_context, instances):
    if session.dirty:
        session.commit()`
```
How can I commit pending changes in the session when a query-invoked auto-flush is done? Also, I don't want to turn off the auto-flush mode

**Comment by zzzeek** [comment]
the Session used to have an "autocommit" mode which we removed for SQLAlchemy 2.0, as it's not considered a great pattern to work in a way that's not transactional.

For modern use, the most recommended pattern to avoid having to call .commit() is to make use of the Session as a context manager and to use .begin():

```py
from sqlalchemy.orm import sessionmaker
session_factory = sessionmaker(engine)

# .. later ...
with session_factory.begin() as sess:
    sess.add(things)

# commits automatically
```

I'm gathering the above is not what you are looking for, however we'd want to ask why would you want autoflush to commit as well.  that's an awkward place for things to commit.

All of that said, if you just dont like transactional atomicity and want statements to commit immediately, turn on driver-level autocommit.   Every INSERT/UPDATE/DELETE will be immediately durable and visible to other transactions: https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#setting-transaction-isolation-levels-dbapi-autocommit

**Reply by liveuser123** [reply]
Thank you for the suggestion to use the session as a context manager with session.begin(). However, as you noted, this approach commits immediately, and we want to maintain transactional atomicity to ensure that all updates happen together.

To clarify, the reason we want to commit any pending changes in the session before a query is invoked is because when a query triggers an auto-flush, it locks the affected rows in a transaction until a manual commit is done. We have multiple places in our codebase where we may forget to commit before a query, which leads to rows being locked for longer than necessary.

To avoid this problem, we want to configure SQLAlchemy so that when a query is invoked and there are dirty changes in the session, it commits those changes before performing the query. This way, we can ensure that rows are not unnecessarily locked and transactions are still atomic.

I hope this clarifies our use case. Let me know if you have any further suggestions or questions.

**Reply by zzzeek** [reply]
> Thank you for the suggestion to use the session as a context manager with session.begin(). However, as you noted, this approach commits immediately, and we want to maintain transactional atomicity to ensure that all updates happen together.
> 
> To clarify, the reason we want to commit any pending changes in the session before a query is invoked is because when a query triggers an auto-flush, it locks the affected rows in a transaction until a manual commit is done. We have multiple places in our codebase where we may forget to commit before a query, which leads to rows being locked for longer than necessary.

There's again a few strategies that I would prefer over embedding a client side COMMIT call into each autoflush.   To be clear, it's not really possible to do the full "session.commit()" call within a session-level event hook.    However since I'm recommending driver-level autocommit in any case, if you want these driver level "autocommits" to occur per-flush rather than per-statement, your event hook can run ".commit()" right on the DBAPI connection, where it will not be detected by SQLAlchemy:

```py
@event.listens_for(Session, "after_flush_postexec")  # run the event *after* the flush...
def commit_pending_changes(session, flush_context):
    # DBAPI-level commit.  SQLAlchemy is none the wiser
    session.connection().connection.driver_connection.commit()
```

However I still would try to solve this issue in other ways first.  The issue with aggressive row locking should be addressed directly if possible.   Particularly if you are using MS SQL Server, you want to [enable snapshot isolation](https://docs.sqlalchemy.org/en/20/dialects/mssql.html#enabling-snapshot-isolation), without this setting SQL server is unusable.  Outside of that issue, I'm not familiar with an overly aggressive row locking behavior that gets in the way to any serious degree unless you are perhaps using SERIALIZABLE isolation.   I would suggest using an appropriate isolation level for those blocks that are sensitive to this issue, as noted at https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#setting-isolation-for-individual-transactions you can set isolation levels for specific transactions if needed.

When you refer to "rows locked for longer than necessary", you also want to keep transactions as short as possible.  if the application has long running transactions, you want to break those up into smaller transactions where the offline time-consuming work is not linked to the lifespan of the transaction itself.

**Reply by liveuser123** [reply]
Thank you for your response and suggestions. Will try the driver level commit approach and see how it works for our use case.

I agree with your point about keeping transactions short and breaking them up into smaller transactions where possible.

Just to clarify, we are actually using Postgres instead of MS SQL Server.

Thank you again for your help.
